apps/src/templates/rubrics/LearningGoals.jsx

```diff
@@ -28,6 +29,149 @@ import ProgressRing from './ProgressRing';
+ * Returns the first and last line that contains the given code snippet.
+ */
+export function findCodeRegion(code, lines, snippet) {
+  // Attempt to just find the code in the full code listing
+  let index = code.indexOf(snippet);
+  let lastIndex = index + snippet.length;
+  if (index < 0) {
+    // Failing to find it, we need to find it line by line instead
+    let context = snippet;
+    let position = 0;
+    for (let line of lines) {
+      // Remember the original length of the line
+      let lineLength = line.length;
+      line = line.trim();
+      if (line !== '') {
+        if (context.startsWith(line)) {
+          if (index < 0) {
+            // Remember the first position in the original code that
+            // the AI references.
+            index = position;
+          }
+          // Remember the last position in the original code.
+          lastIndex = position + lineLength;
+          context = context.substring(line.length + 1).trim();
+          if (context === '') {
+            // All of the code was found
+            break;
+          }
+        } else {
+          // We didn't match it. Reset our search.
+          context = snippet;
+          index = -1;
+        }
+      }
+
+      // Move the current search position
+      position += lineLength + 1;
+    }
+  }
+
+  // As long as we found a region, we will determine what the line
+  // number was for the first and last line of the region.
+  if (index < 0) {
+    return [null, null];
+  }
+
+  let lineNumber = (code.substring(0, index).match(/\n/g) || []).length + 1;
+  let lastLineNumber =
+    (code.substring(0, lastIndex).match(/\n/g) || []).length + 1;
+
+  return [lineNumber, lastLineNumber];
+}
+
+/**
+ * Adds annotations to the source code viewed in the current editor based
+ * on the AI observations of the learning goal referenced by the given index.
+ */
+export function annotateLines(observations) {
+  // Get a reference to all the whitespaced lines of code
+  const code = getAnonymizedCode();
+  const lines = code.split('\n');
+
+  // Go through the AI observations
+  // For every reference the AI gave us, we will find it in the code.
+  // The AI has trouble giving line numbers, so even though we parse
+  // those out, we do not trust them and instead find the code it
+  // references to highlight it.
+  for (const match of observations.matchAll(
+    'Line (\\d+)(?:\\s*-\\s*(\\d+))?:(.+?)\\s*(?=Line|$)'
+  )) {
+    let lineNumber = parseInt(match[1]);
+    let lastLineNumber = parseInt(match[2] || lineNumber);
+    let found = false;
+    const context = match[3].trim();
+
+    const message = context
+      .substring(0, context.indexOf('`') || context.length)
+      .trim();
+
+    // We look at all of the code references the AI gave us which are
+    // surrounded by backticks.
+    const references = context.substring(message.length);
+    for (const submatch of references.matchAll(/`([^\`]+)`/g)) {
+      let snippet = submatch[1].trim();
```

Comment:
if the AI provided something like this:
```
Line 1: the code is ` ` 
```
could we bail here if the snippet is empty, and cover with a unit test? not sure how likely this is, but seems worth covering as long as it is straightforward to add

Reply:
That makes sense. That's a good worst case... It would annotate the first line most likely and be a bit confusing... makes sense to ignore it.

Reply:
This was a very good catch. I really appreciate you noticing this. In the case where there would be an empty string marked by the AI as the referenced code, the matcher would be looking for an empty string. It would immediately "find" an empty string and report it as line 1.

I have added logic to bail if the snippet is empty in the `annotateLines` function and, as an extra precaution, added logic to return 'not found' from `findCodeRegion` when the search string is empty/whitespace.

Then, I have it more or less throw out the annotation since it didn't provide any code as evidence. Not sure if we should just annotate via the line it gives.

Reply:
yay, glad this was helpful! 

> Then, I have it more or less throw out the annotation since it didn't provide any code as evidence. Not sure if we should just annotate via the line it gives.

do you think there will be other helpful info in the annotation that make it worth surfacing, like:
```
Line 3: the user demonstrates good use of whitespace: `  `
```
since this is an early exploration, I think there is some leeway to surface this info even if we might want to tighten down on what we show to users later.

Just my 2 cents, happy for you to decide here.